cirq/sim/clifford_simulator.py

```diff
@@ -0,0 +1,749 @@
+
+    def _X(self,q):
+        self._H(q)
+        self._Z(q)
+        self._H(q)
+
+    def _Y(self,q):
+        self._Z(q)
+        self._X(q)
+        self.omega *= 1j
+            
+    def _CZ(self,q,r,right=False):
+        if right:
+            self.M[:,q] ^= self.F[:,r] 
+            self.M[:,r] ^= self.F[:,q] 
+            self.gamma[:] = (self.gamma[:] + 2*self.F[:,q]*self.F[:,r]) % 4
+        else:
+            self.M[q,:] ^= self.G[r,:] 
+            self.M[r,:] ^= self.G[q,:] 
+            
+    def _CNOT(self,q,r,right=False):
+        if right:
+            self.G[:,q] ^= self.G[:,r] 
+            self.F[:,r] ^= self.F[:,q] 
+            self.M[:,q] ^= self.M[:,r] 
+        else:
+            self.gamma[q] = (self.gamma[q] + self.gamma[r] + 2*(sum(self.M[q,:]&self.F[r,:])%2)) % 4
+            self.G[r,:] ^= self.G[q,:] 
+            self.F[q,:] ^= self.F[r,:] 
+            self.M[q,:] ^= self.M[r,:] 
+            
+    def _H(self,p):
+        t = self.s ^ (self.G[p,:] & self.v)
+        u = self.s ^ (self.F[p,:] & (~self.v)) ^ (self.M[p,:] & self.v)
+        
+        alpha = sum(self.G[p,:] & (~self.v) & self.s) % 2
+        beta = sum(self.M[p,:] & (~self.v) & self.s) 
+        beta += sum(self.F[p,:] & self.v & self.M[p,:])
+        beta += sum(self.F[p,:] & self.v & self.s)
+        beta %= 2
+        
+        delta = (self.gamma[p] + 2*(alpha+beta)) % 4
+        
+        self._update_sum(t,u,delta=delta,alpha=alpha)
+    
+    def _update_sum(self,t,u,delta=0,alpha=0):
+        ''' Implements the transformation (Proposition 4 in Bravyi et al)
+
+                i^alpha U_H (|t> + i^delta |u>) = omega W_C W_H |s'>
+        '''
+        if np.all(t == u):
+            self.s = t
+            self.omega *= 1/np.sqrt(2)*(-1)**alpha*(1+1j**delta)
+        else:
+            set0 = np.where((~self.v) & (t ^ u))[0]
+            set1 = np.where(self.v & (t ^ u))[0]
+            
+            # implement Vc
+            if len(set0) > 0:
+                q = set0[0]
+                for i in set0:
+                    if i != q:
+                        self._CNOT(q,i,right=True)
+                for i in set1:
+                    self._CZ(q,i,right=True)
+            elif len(set1) > 0:
+                q = set1[0]
+                for i in set1:
+                    if i != q:
+                        self._CNOT(i,q,right=True)
+                
+            e = np.zeros(self.n,dtype=bool)
+            e[q] = True
+            
+            if t[q]:
+                y = u ^ e
+                z = u
+            else:
+                y = t
+                z = t ^ e
+                
+            (omega,a,b,c) = self._H_decompose(self.v[q],y[q],z[q],delta)
+            
+            self.s = y
+            self.s[q] = c
+            self.omega *= (-1)**alpha*omega
+            
+            if a:
+                self._S(q,right=True)
+            self.v[q] ^= b ^ self.v[q]
+
+    def _H_decompose(self,v,y,z,delta):
```

Comment:
This method doesn't use self; it can be moved out of the class

Reply:
We have been defining methods that don't use `self` everywhere in the codebase. How is this the exception?

PS: There is a `no-self-use` setting in pylint that catches this. We could turn it on and fix all of them.